posthog/special_migrations/utils.py

```diff
@@ -0,0 +1,150 @@
+
+    attempt_migration_rollback(migration_instance)
+
+
+def trigger_migration(migration_instance: SpecialMigration, fresh_start=True):
+    from posthog.tasks.special_migrations import run_special_migration
+
+    task = run_special_migration.delay(migration_instance.name, fresh_start)
+
+    update_special_migration(
+        migration_instance=migration_instance, celery_task_id=str(task.id),
+    )
+
+
+def force_stop_migration(migration_instance: SpecialMigration, error: str = "Force stopped by user"):
+    """
+    In theory this is dangerous, as it can cause another task to be lost 
+    `revoke` with `terminate=True` kills the process that's working on the task
+    and there's no guarantee the task will not already be done by the time this happens.
+    See: https://docs.celeryproject.org/en/stable/reference/celery.app.control.html#celery.app.control.Control.revoke
+    However, this is generally ok for us because:
+    1. Given these are long-running migrations, it is statistically unlikely it will complete during in between 
+    this call and the time the process is killed
+    2. Our Celery tasks are not essential for the functioning of PostHog, meaning losing a task is not the end of the world
+    """
+
+    app.control.revoke(migration_instance.celery_task_id, terminate=True)
+    process_error(migration_instance, error)
+
+
+def force_rollback_migration(migration_instance: SpecialMigration):
+    from posthog.special_migrations.runner import attempt_migration_rollback
+
+    attempt_migration_rollback(migration_instance, force=True)
+
+
+def complete_migration(migration_instance: SpecialMigration):
+    update_special_migration(
+        migration_instance=migration_instance,
+        status=MigrationStatus.CompletedSuccessfully,
+        finished_at=datetime.now(),
+        progress=100,
+        lock_row=True,
+    )
+
+    from posthog.special_migrations.runner import run_next_migration
+
+    next_migration = DEPENDENCY_TO_SPECIAL_MIGRATION.get(migration_instance.name)
+
+    if next_migration:
+        run_next_migration(next_migration)
+
+
+def reset_special_migration(migration_instance: SpecialMigration):
+    update_special_migration(
+        migration_instance=migration_instance,
+        last_error="",
+        current_query_id="",
+        celery_task_id="",
+        progress=0,
+        current_operation_index=0,
+        status=MigrationStatus.Running,
+        started_at=datetime.now(),
+        finished_at=None,
+    )
+
+
+def update_special_migration(
+    migration_instance: SpecialMigration,
+    last_error: Optional[str] = None,
+    current_query_id: Optional[str] = None,
+    celery_task_id: Optional[str] = None,
+    progress: Optional[int] = None,
+    current_operation_index: Optional[int] = None,
+    status: Optional[int] = None,
+    started_at: Optional[datetime] = None,
+    finished_at: Optional[datetime] = None,
+    lock_row=False,
```

Comment:
maybe safer to default to true here & explicitly set false if it's cool

Reply:
potentially - but `lock_row=True` allows me to more easily find references to the updates that can happen simultaneously though, which prevents errors

Reply:
Yeah seems like it might be a good idea to do this with the row lock each time. Would be pretty sketchy if you had this get inconsistent.